### ISSUE TITLE
Consider adding a logger pool

### ISSUE THREAD

**Issue by josevalim** [elixir-lang/elixir#5324]
Instead of having a single logger process that may be a processing bottleneck, we could allow a pool of loggers to be started and dispatch to it based on the caller PID. When a handler is added, it should be added to all logger pool entries.

/cc @fishcakez

**Comment by k-solutions** [comment]
Yes I was thinking in that direction too and we should add some : logger_pool_size application config option.
How about non local PID logging?

**Comment by eidge** [comment]
Is this something you would like to see someone prototype, or just discuss at the moment?

Also, are you keen on keeping the GenEvent or should this wait for GenStage to be stable?

(I would love  to help with this if possible)

**Comment by josevalim** [comment]
We certainly would not migrate to GenStage. It would be a pool of GenStage
and we would take care of adding handlers to all pool entries.

We may get rid of GenEvent too but we are not yet sure if it is worth it.
Probably not if we introduce a pool.


-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by eidge** [comment]
Yup, makes sense. Mind if I take a stab at this? Or is it not up for grabs yet?

**Comment by josevalim** [comment]
Not for grabs yet because we still need to benchmark and test multiple scenarios. If you want to start with benchmarks, that would be welcome. We need to set up benchmarks for the current solutions and use it to verify prototypes.

**Comment by eidge** [comment]
Ok makes sense.

Will create a sample app to benchmark the Logger here: https://github.com/eidge/elixir_logger_benchmark

**Comment by eidge** [comment]
I've done an initial benchmark, for the console backend a noop backend and a really slow backend.

TLDR: On my machine (VM with 2 cores running inside a i7 @ 2.5 ghz Macbook Pro)

```
Console backend: 8700 logs/s
Noop backend: 140 000 logs/s
Slow backend: +infinity
```

This is what I think is going on:

Basically we've got two bottlenecks the first one is the speed of the backend in itself, since we will default to synchronous calls once the mailbox reaches a configurable number of queued messages. We can see this by the fact that the Noop backend will run much faster than the console backend.

The second bottleneck is the capability of the Logger (GenEvent) to dispatch messages, which I believe is the cap we see on the Noop backend.

Having a process pool would multiply the throuput of the Logger, but for the console backend we would be capped by the IO throuput anyway, since the backend itself will start doing sync calls if there are too much logs queued (intended back-pressure) and the pool wouldn't help much here after we saturate the logger.

Repo here: https://github.com/eidge/elixir_logger_benchmark

**Comment by josevalim** [comment]
@eidge great work, thank you.

What if we also benchmark how long it would take for those operations by wrapping directly to the IO device by using something such as `IO.puts(:stdio, log)`? Maybe we can then find information on how much we can saturate the IO process?

**Comment by eidge** [comment]
Thanks :)

I've added a **ConsolePutsBackend** that does everything the core :console does, except it uses IO.puts for IO (I wanted to see if :console was slow, it isn't an huge performance impact).

And a PutsBackend that does nothing except put the ```{Logger, msg, _, _}``` msg.

```
ConsolePutsBackend: 14500 logs/s
PutsBackend: 17000 logs/s
```

Curiously enough, the benchmark runs faster if I switch my tmux window to a different one while it is runnning. Wonder if we can improve the performance by sending logs in batches rather than one by one.

**Comment by josevalim** [comment]
> running mix run bench/console_puts_backend.exs > ./bench.txt

And how does that affect the regular console backend?

**Comment by eidge** [comment]
When redirecting output to file:

```
Console Backend: 96 000 logs/s
Noop Backend: 140 000 logs/s (no change as expected)
Console Puts Backend: 37 000 logs/s
Puts Backend: 37 000 log/s
```

**Comment by josevalim** [comment]
@eidge great!

I have pushed a very simple commit that starts 3 other Logger genevent, simulating the pool: https://github.com/elixir-lang/elixir/tree/jv-logger-sample

Do we see any difference for the console backend, with and without redirecting to a file, with this very small proof of concept?

If we see a big difference, then it would justify a pool, otherwise it seems console works quite well as long as we are redirecting to a file.

**Comment by eidge** [comment]
With the Elixir from that commit, things actually get a bit slower (probably within std dev though)

```
Single Logger
Console Backend: 8700 logs/s
Console Backend (> file): 96 000 logs/s

4 Logger Processes
Console Backend 8400 logs/s
Console Backend (> file): 92 000 logs/s
```

The main bottleneck seems to be the backend and not the logger, the logger seems to cap (on my machine) at 140k logs/s. This means that unless someone would actually need to log more than 140k logs/s, they would be able to by implementing a fast logger backend.

The Noop backend got faster in my machine just by a small amount (5%) (2 cores only, so about half is being consumed by the tasks that are actually generating the logs), but I expected a much bigger improvement. Taking a look at top shows that I'm saturating my two cores, so it would probably be worth it to run this on a bigger machine.

I do have one question, though, if we do implement a logger pool don't we risk logs out of order?

**Comment by josevalim** [comment]
We would send messages based on the caller PID, so we lose order but
messages from the same process are still ordered. However, everything
points to not adding a pool. :)

-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by eidge** [comment]
Yup that makes sense. :)

**Comment by josevalim** [comment]
Alright, I will close this issue for now then. There is still a chance the GenEvent will be a bottleneck on machines with a high number of cores but we can wait until we have our hands on one of those.

Thank you @eidge! Your contributions were very valuable!  :heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
